src/node_modules/app/components/elementary/ElasticSearchField.svelte

```diff
@@ -0,0 +1,177 @@
+		string: 'text'
+	}
+
+	function pad (num, size) {
+		return `${num}`.padStart(size, '0');
+	}
+
+	function Date_YYYYMMDD_dash (date) {
+		return `${date.getFullYear()}-${pad(date.getMonth() + 1, 2)}-${pad(date.getDate(), 2)}`;
+	}
+
+	function Date_YYYYMMDD_dash_time (date) {
+		return `${Date_YYYYMMDD_dash(date)} ${pad(date.getHours(), 2)}:${pad(date.getMinutes(), 2)}:${pad(date.getSeconds(), 2)}`;
+	}
+
+	function incomingBoolean (incoming) {
+		if (incoming === null || incoming === undefined) {
+			return 'null';
+		}
+		return Boolean(incoming);
+	}
+	function outgoingBoolean (outgoing) {
+		if (outgoing === 'null') {
+			return null;
+		}
+		return outgoing;
+	}
+
+	const incomingFormatters = {
+		'boolean': incomingBoolean,
+		'Opaque<number, "integer">': v => v,
+		'Opaque<number, "float">': v => v,
+		'Opaque<string, "Date_YYYYMMDD_dash">': dateString => Date_YYYYMMDD_dash(new Date(dateString)),
+		'Opaque<string, "Date_YYYYMMDD_dash_time">': dateString => Date_YYYYMMDD_dash(new Date(dateString)),
+	}
+
+	const outgoingFormatters = {
+		'Opaque<string, "Date_YYYYMMDD_dash">': dateString => Date_YYYYMMDD_dash(new Date(dateString)),
+		'Opaque<string, "Date_YYYYMMDD_dash_time">': dateString => Date_YYYYMMDD_dash_time(new Date(dateString)),
+		'Opaque<number, "integer">': parseInt,
+		'Opaque<number, "float">': parseFloat,
+	}
+
+	const placeholders = {
+		'Opaque<number, "integer">': 'integer',
+		'Opaque<number, "float">': 'float',
+		'Opaque<string, "ES_keyword">': 'text',
+		'Opaque<string, "ES_text">': 'text',
+		'Opaque<string, "ES_text_w_keyword">': 'text',
+	}
+
+	function computeInputType (esType) {
+		return inputTypes[esType];
+	}
+
+	function formatIncomingValue (incomingValue) {
+		let result;
+		switch (dataType) {
+			case 'boolean':
+				result = incomingBoolean(incomingValue);
+				break;
+			default:
+				if (dataType in incomingFormatters && incomingValue !== null && incomingValue !== undefined) {
+					result = incomingFormatters[dataType](incomingValue)
+				}
+				else {
+					result = incomingValue || '';
+				}
+		}
+		return result;
+	}
+
+	function notify (payload) {
+		let outgoing = payload;
+		if (dataType === 'boolean') {
+			outgoing = outgoingBoolean(payload);
+		}
+		else if (dataType in outgoingFormatters) {
+			outgoing = outgoing.length > 0
+				? outgoingFormatters[dataType](outgoing)
+				: null;
+		}
+		dispatch('change', outgoing);
+	}
+
+	function handleChange (event) {
+		notify(event.target.value);
+	}
+
+	function handleInput (event) {
+		currentValue = event.target.value;
+		// TODO Perhaps we need to handle validation as well
+	}
+
+	function computePlaceHolder (newDataType) {
+		if (newDataType in placeholders) {
+			return placeholders[newDataType];
+		}
+		return newDataType;
+	}
+
+	$: placeholder = computePlaceHolder(dataType);
+	$: {
+		fieldValue = formatIncomingValue(value);
+		currentValue = fieldValue;
+	}
+	$: type = computeInputType(dataType);
```

Comment:
not sure we need to invoke a function here, I'd just use

`$: placeholder = inputTypes[esType];`